src/Type/Php/PregMatchFunctionTypeSpecifyingExtension.php

```diff
@@ -0,0 +1,114 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Analyser\SpecifiedTypes;
+use PHPStan\Analyser\TypeSpecifier;
+use PHPStan\Analyser\TypeSpecifierAwareExtension;
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\BitwiseFlagHelper;
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\FunctionTypeSpecifyingExtension;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\TypeCombinator;
+use function strtolower;
+
+final class PregMatchFunctionTypeSpecifyingExtension implements FunctionTypeSpecifyingExtension, TypeSpecifierAwareExtension
+{
+
+	public function __construct(
+		private BitwiseFlagHelper $bitwiseFlagAnalyser,
+	)
+	{
+	}
+
+	private TypeSpecifier $typeSpecifier;
+
+	public function isFunctionSupported(
+		FunctionReflection $functionReflection,
+		FuncCall $node,
+		TypeSpecifierContext $context,
+	): bool
+	{
+		return strtolower($functionReflection->getName()) === 'preg_match'
+			&& !$context->null();
+	}
+
+	public function specifyTypes(
+		FunctionReflection $functionReflection,
+		FuncCall $node,
+		Scope $scope,
+		TypeSpecifierContext $context,
+	): SpecifiedTypes
+	{
+		$args = $node->getArgs();
+		$matchesArg = $args[2]->value ?? null;
+		$optionsExpr = $args[3]->value ?? null;
+
+		if ($matchesArg === null) {
+			return new SpecifiedTypes();
+		}
+
+		if ($context->false()) {
+			return $this->typeSpecifier->create(
+				$matchesArg,
+				new ArrayType(new MixedType(), new MixedType()),
+				$context->negate(),
+				false,
+				$scope,
+			);
+		}
+
+		$valueType = new StringType();
+		if ($optionsExpr !== null) {
+			$unmatchedAsNull = $this->bitwiseFlagAnalyser->bitwiseOrContainsConstant($optionsExpr, $scope, 'PREG_UNMATCHED_AS_NULL')->yes();
+			$offsetCapture = $this->bitwiseFlagAnalyser->bitwiseOrContainsConstant($optionsExpr, $scope, 'PREG_OFFSET_CAPTURE')->yes();
```

Comment:
Just curious, will this correctly handle code like https://github.com/composer/pcre/blob/main/src/Preg.php#L29-L33 that explicitly ensures that a flag is not part of $flags? I guess running this PR on composer/pcre would be interesting to see if you get any errors.

Reply:
> ensures that a flag is not part of $flag

Sounds like a good example what the newly added BitwiseFlagHelper could learn